File: Trees/Binarytree.js
class Node
{
    constructor(data)
    {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}


export default class BinaryTree{
  constructor(root){
    this.root=null
 
  }


insert(data)
{
    
    let newNode = new Node(data);

    if(this.root === null)
        this.root = newNode;
    else
 
        this.insertNode(this.root, newNode);
}
insertNode(node, newNode)
{
    if(newNode.data < node.data)
    {
        if(node.left === null)
            node.left = newNode;
        else
 
         
            this.insertNode(node.left, newNode);
    }
 
    else
    {
        if(node.right === null)
            node.right = newNode;
        else
 
            
            this.insertNode(node.right,newNode);
    }
}

MinNode(node)
{
   
    if(node.left === null)
        return node;
    else
        return this.MinNode(node.left);
}

delete(data)
{
    this.root = this.deleteNode(this.root, data);
}
 
deleteNode(node, key)
{
         
    if(node === null)
        return null;
 
    
    else if(key < node.data)
    {
        node.left = this.deleteNode(node.left, key);
        return node;
    }
 
    
    else if(key > node.data)
    {
        node.right = this.deleteNode(node.right, key);
        return node;
    }
 
    else
    {
         if(node.left === null && node.right === null)
        {
            node = null;
            return node;
        }
 
        
        if(node.left === null)
        {
            node = node.right;
            return node;
        }
         
        else if(node.right === null)
        {
            node = node.left;
            return node;
        }
 
       
        let temp = this.MinNode(node.right);
        node.data = temp.data;
 
        node.right = this.deleteNode(node.right, temp.data);
        return node;
    }
 
}

}
let test= new BinaryTree()
test.insert(7)
test.insert(4)
test.insert(9)
test.delete(4)
console.log(test)/*BinaryTree {
    root: Node {
      data: 7,
      left: null,
      right: Node { data: 9, left: null, right: null }
    }*/

File: Non-Linear Data Structure/Trees/BinaryTree.py
class Node:
    def __init__(self,data):
        self.value=data
        self.left=None
        self.right=None

class Tree:
    def __init__(self) -> None:
        self.root=None
        self.size=0

    def insert(self,data):
      if self.root==None:
            self.root=Node(data)
            self.size+=1
            
      else:
       self.help_insert(data,self.root)


    def help_insert(self,data,root):
         if root.value<data:
            if root.right is None:
                root.right=Node(data)
                self.size+=1
            else:
             self.help_insert(data,root.right)
        
         if root.value>data:
            if root.left is None:
                root.left=Node(data)
                self.size+=1
            else:
             self.help_insert(data,root.left)


    def getNode(self):
        if self.root is None:
            return

    
    def minNode(self,node):
       if(not node.left):
          return node
       else:
          self.minNode(self,node.left)
    
    def delete(self,value):
       self.root=self.__help_delete(self.root,value)

    def __help_delete(self,root,value):
       if(root.value==value):
          if(root.right==None and root.left):
             root=root.left 
             return root
          elif(root.right and root.left==None):
             root=root.right
             return root
          
        #If both children are not null
          elif(root.right and root.left):
             sub=self.minNode(root.right)
             root.value=sub.value
             root.right=self.__help_delete(root.right,sub.value)
             return root
          
      #If both children are null
          elif(not root.right and not root.left):
             root=None
             return root
      
      #Traverse through the right subtreee
       elif(root.value< value):
          root.right=self.__help_delete(root.right,value)
          return root 
       
      #Traverse through the left subtree
       elif(root.value> value):
          root.left=self.__help_delete(root.left,value)
          return root    

    
        

test=Tree()
test.insert(3)
test.insert(8)
test.insert(4)

test.delete(8)
print(test.root.right.value)